### PR TITLE
when: emit switch for String if possible

### PR DESCRIPTION
**Pull request by ting-yuan**
Effectively, the following when structure:
```
  when (s) {
    s1, s2 -> e1,
    s3 -> e2,
    s4 -> e3,
    ...
    else -> e
  }
```
is implemented as:
```
  when (s.hashCode()) {
    h1 -> {
      if (s == s1)
        e1
      else if (s == s2)
        e1
      else if (s == s3)
        e2
      else
        e
    }
    h2 -> if (s == s3) e2 else e,
    ...
    else -> e
  }
```
where `s1.hashCode() == s2.hashCode() == s3.hashCode() == h1,
      s4.hashCode() == h2`.

A `tableswitch` or `lookupswitch` is used for the hash code lookup.

### CODE REVIEW

File: compiler/testData/codegen/bytecodeText/whenStringOptimization/duplicatingItemsSameHashCode3.kt [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+fun foo(x : String) : String {
+    assert("abz]".hashCode() == "aby|".hashCode())
+
+    when (x) {
+        "abz]" -> return "abz"
+        "ghi"  -> return "ghi"
+        "aby|" -> return "aby"
+        "abz]" -> return "fail"
+        "xyz" -> return "xyz"
```
Comment (max-kammerer) [inline compiler/testData/codegen/bytecodeText/whenStringOptimization/duplicatingItemsSameHashCode3.kt]: Please add box test for this case similar to original one in compiler/testData/codegen/box/when/stringOptimization/duplicatingItemsSameHashCode.kt

File: compiler/testData/codegen/bytecodeText/whenStringOptimization/duplicatingItemsSameHashCode2.kt [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+        "abz]" -> return "abz"
+        "ghi"  -> return "ghi"
+        "aby|" -> return "aby"
+        "abz]" -> return "fail"
+    }
+
+    return "other"
+}
+
+// 0 LOOKUPSWITCH
```
Comment (max-kammerer) [inline compiler/testData/codegen/bytecodeText/whenStringOptimization/duplicatingItemsSameHashCode2.kt]: Is LOOKUPSWITCH is absent cause of hashAndSwitchLabels.size <= 2 condition?

Comment (ting-yuan) [reply compiler/testData/codegen/bytecodeText/whenStringOptimization/duplicatingItemsSameHashCode2.kt]: Yes. I've updated the comments in it.

Comment (max-kammerer) [reply compiler/testData/codegen/bytecodeText/whenStringOptimization/duplicatingItemsSameHashCode2.kt]: Please also add box test for such case

Comment (ting-yuan) [reply compiler/testData/codegen/bytecodeText/whenStringOptimization/duplicatingItemsSameHashCode2.kt]: Didn't add it because it is exactly the same to the existing test. Will update after the tests finished.

File: compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/SwitchGenerator.kt [hunk 3]
```diff
@@ -137,58 +173,190 @@ class SwitchGenerator(private val expression: IrWhen, private val data: BlockInf
...
+
+        protected abstract fun genSubject()
+
+        protected abstract fun genSwitch()
+
+        protected fun genIntSwitch(unsortedIntCases: List<Pair<Int, Label>>) {
+            val intCases = unsortedIntCases.sortedBy { it.first }
+            val caseMin = intCases.first().first
+            val caseMax = intCases.last().first
+            val rangeLength = caseMax - caseMin + 1L
```
Comment (max-kammerer) [inline compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/SwitchGenerator.kt]: val rangeLength = caseMax.toLong() - caseMin.toLong() + 1L
It would be best to also add some test for overflow case

File: compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/SwitchGenerator.kt [hunk 4]
```diff
@@ -137,58 +173,190 @@ class SwitchGenerator(private val expression: IrWhen, private val data: BlockInf
...
-    private fun gen(
-        cases: ArrayList<Pair<Int, Label>>,
+    abstract inner class Switch(
+        val subject: IrGetValue,
+        val defaultLabel: Label,
+        val endLabel: Label,
+        val elseExpression: IrExpression?,
+        val thenExpressions: ArrayList<Pair<IrExpression, Label>>
+    ) {
+        open fun gen(): StackValue? {
```
Comment (max-kammerer) [inline compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/SwitchGenerator.kt]: I would prefer to explicitly clarify here that gen works only if there are enough cases
gen -> genOptimizedIfEnoughCases (not sure that new name couldn't be better)?

File: compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/SwitchGenerator.kt [hunk 5]
```diff
@@ -43,36 +43,72 @@ class SwitchGenerator(private val expression: IrWhen, private val data: BlockInf
...
+        // switch isn't applicable if there's no case at all, e.g., when() { else -> ... }
+        if (allConditions.size == 0)
             return null
 
-        if (areConstIntComparisons(allConditions.map { it.first })) {
+        // Remove labels and "then expressions" that are not reachable.
+        fun removeUnreachableExpressions(reachables: HashSet<Label>) = thenExpressions.removeIf { it.second !in reachables }
+
+        val calls = allConditions.map { it.first }
+        if (areConstIntComparisons(calls)) {
```
Comment (max-kammerer) [inline compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/SwitchGenerator.kt]: Branches seem identical except return value: maybe merge then with explicit cast for cases in each constructor

File: compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/SwitchGenerator.kt [hunk 6]
```diff
@@ -43,36 +43,72 @@ class SwitchGenerator(private val expression: IrWhen, private val data: BlockInf
...
             // if all conditions are CALL EQEQ(tmp_variable, some_int_constant)
-            val cases = allConditions.mapTo(ArrayList()) { Pair((it.first.getValueArgument(1) as IrConst<*>).value as Int, it.second) }
             val subject = allConditions[0].first.getValueArgument(0)!! as IrGetValue
-            return gen(cases, subject, defaultLabel, endLabel, elseExpression, thenExpressions)
+
+            // Don't generate repeated cases, which are unreachable but allowed in Kotlin.
+            // Only keep the first encountered case:
+            val cases =
+                allConditions.map { Pair((it.first.getValueArgument(1) as IrConst<*>).value as Int, it.second) }.distinctBy { it.first }
```
Comment (max-kammerer) [inline compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/SwitchGenerator.kt]: Please use data class instead of Pair

File: compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/SwitchGenerator.kt [hunk 7]
```diff
@@ -43,36 +43,72 @@ class SwitchGenerator(private val expression: IrWhen, private val data: BlockInf
...
 
-        // IF is more compact when there are only 1 or fewer branches, in addition to else.
-        if (allConditions.size <= 1)
+        // switch isn't applicable if there's no case at all, e.g., when() { else -> ... }
+        if (allConditions.size == 0)
             return null
 
-        if (areConstIntComparisons(allConditions.map { it.first })) {
+        // Remove labels and "then expressions" that are not reachable.
+        fun removeUnreachableExpressions(reachables: HashSet<Label>) = thenExpressions.removeIf { it.second !in reachables }
```
Comment (max-kammerer) [inline compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/SwitchGenerator.kt]: Please add box test for integers (if it's not present)

File: compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/SwitchGenerator.kt [hunk 8]
```diff
@@ -137,58 +173,190 @@ class SwitchGenerator(private val expression: IrWhen, private val data: BlockInf
...
+        subject: IrGetValue,
+        defaultLabel: Label,
+        endLabel: Label,
+        elseExpression: IrExpression?,
+        thenExpressions: ArrayList<Pair<IrExpression, Label>>,
+        private val cases: List<Pair<String?, Label>>
+    ) : Switch(subject, defaultLabel, endLabel, elseExpression, thenExpressions) {
+
+        private val hashToStringAndExprLabels = HashMap<Int, ArrayList<Pair<String?, Label>>>()
+        private val hashAndSwitchLabels = ArrayList<Pair<Int, Label>>()
```
Comment (max-kammerer) [inline compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/SwitchGenerator.kt]: Avoid Pair usage (it's not trivial to read code with pairs below)

File: compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/SwitchGenerator.kt [hunk 9]
```diff
@@ -137,58 +173,190 @@ class SwitchGenerator(private val expression: IrWhen, private val data: BlockInf
...
-            mv.lookupswitch(defaultLabel, cases.map { it.first }.toIntArray(), cases.map { it.second }.toTypedArray())
-        } else {
-            val labels = Array(rangeLength.toInt()) { defaultLabel }
+        thenExpressions: ArrayList<Pair<IrExpression, Label>>,
+        private val cases: List<Pair<Int, Label>>
+    ) : Switch(subject, defaultLabel, endLabel, elseExpression, thenExpressions) {
+
+        override fun gen(): StackValue? {
+            // IF is more compact when there are only 1 or fewer branches, in addition to else.
+            if (cases.size <= 1)
```
Comment (max-kammerer) [inline compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/SwitchGenerator.kt]: Could  be extracted to precondition check and called within base class

File: compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/SwitchGenerator.kt (hunk 5 above)
Comment (ting-yuan) [reply compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/SwitchGenerator.kt]: Done. `areConstComparisons` are hoisted out of `areConst[type]Comparisons` to guard the shared operations, which are moved before `areConst[type]Comparisons`.

File: compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/SwitchGenerator.kt [hunk 10]
```diff
@@ -38,41 +42,82 @@ class SwitchGenerator(private val expression: IrWhen, private val data: BlockInf
...
+        // Don't generate repeated cases, which are unreachable but allowed in Kotlin.
+        // Only keep the first encountered case:
+        val cases =
+            callToLabels.map { ValueToLabel((it.call.getValueArgument(1) as IrConst<*>).value, it.label) }.distinctBy { it.value }
+
+        // Remove labels and "then expressions" that are not reachable.
+        val reachableLabels = HashSet(cases.map { it.label })
+        expressionToLabels.removeIf { it.label !in reachableLabels }
+
+        if (areConstIntComparisons(calls)) {
```
Comment (max-kammerer) [inline compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/SwitchGenerator.kt]: Maybe best
```
return when {
            areConstIntComparisons(calls) ->
                IntSwitch(
                    subject,
                    defaultLabel,
                    endLabel,
                    elseExpression,
                    expressionToLabels,
                    cases
                )
            areConstStringComparisons(calls) ->
                StringSwitch(
                    subject,
                    defaultLabel,
                    endLabel,
                    elseExpression,
                    expressionToLabels,
                    cases
                )
            else -> null // TODO: Enum, etc.
        }?.genOptimizedIfEnoughCases()
```

File: compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/SwitchGenerator.kt [hunk 11]
```diff
@@ -137,58 +182,184 @@ class SwitchGenerator(private val expression: IrWhen, private val data: BlockInf
...
-        val result = if (elseExpression == null) {
-            // There's no else part. No stack value will be generated.
-            StackValue.putUnitInstance(mv)
-            onStack(Type.VOID_TYPE)
-        } else {
-            // Generate the else part.
-            mv.visitLabel(defaultLabel)
-            val stackValue = elseExpression.run { gen(this, data) }
-            coerceNotToUnit(stackValue.type, stackValue.kotlinType, expression.type.toKotlinType())
+        override fun shouldOptimize() = hashAndSwitchLabels.size > 2
```
Comment (max-kammerer) [inline compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/SwitchGenerator.kt]: Please add comment to this heuristic (seems it was lost)

File: compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/SwitchGenerator.kt (hunk 10 above)
Comment (ting-yuan) [reply compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/SwitchGenerator.kt]: Looks better, thanks!

**Comment by max-kammerer** [comment]
@ting-yuan Many Thanks!
